File: elixir/custom-set/custom_set.exs
# I could use a `Map` whose keys are the values in the `Set`, but that feels
# like cheating because it'd make everything so easy.
defmodule CustomSet do
  # This lets the compiler check that all Set callback functions have been
  # implemented.
  @behaviour Set

  defstruct list: []

  def new() do
    %CustomSet{}
  end

  def new(coll) do
    %CustomSet{list: coll |> Enum.uniq |> Enum.sort}
  end

  def empty(_cs), do: new

  def delete(cs, value) do
    new(cs.list |> List.delete(value))
  end

  def difference(cs1, cs2) do
    new(cs1.list -- cs2.list)
  end

  def disjoint?(cs1, cs2) do
    # We could make this faster by short-circuiting, but with the sizes
    # we're working with it doesn't matter.
    same = for v1 <- cs1.list,
               v2 <- cs2.list,
               v1 === v2 do
                 v1
               end
    length(same) == 0
  end

  def equal?(%CustomSet{list: s1}, %CustomSet{list: s2}) when length(s1) != length(s2), do: false
  def equal?(%CustomSet{list: s1}, %CustomSet{list: s2}) do
    Enum.zip(Enum.sort(s1), Enum.sort(s2))
    |> Enum.reduce(true, fn({v1, v2}, acc) ->
      if v1 != v2, do: false, else: acc
    end)
  end

  def intersection(cs1, cs2) do
    same = for v1 <- cs1.list,
               v2 <- cs2.list,
               v1 === v2 do
                 v1
               end
    new(same)
  end

  def member?(cs, value) do
    length(cs.list) != length(cs.list -- [value])
  end

  def put(cs, value) do
    new([value | cs.list] |> Enum.sort)
  end

  def size(cs), do: length(cs.list)

  def subset?(cs1, cs2) do
    length(cs1.list -- cs2.list) == 0
  end

  def to_list(cs), do: cs.list

  def union(cs1, cs2), do: new(cs1.list ++ cs2.list)
end

File: go/custom-set/custom_set.go
package stringset

import (
	"fmt"
	"sort"
	"strings"
)

const testVersion = 3

type Set struct {
	set map[string]bool
}

func New() Set {
	return Set{set: make(map[string]bool)}
}

func NewFromSlice(strs []string) Set {
	set := New()
	for _, s := range strs {
		set.set[s] = true
	}
	return set
}

func (set Set) Add(s string) {
	set.set[s] = true
}

func (set Set) Delete(s string) {
	delete(set.set, s)
}

func (set Set) Has(s string) bool {
	return set.set[s]
}

func (set Set) IsEmpty() bool {
	return len(set.set) == 0
}

func (set Set) Len() int {
	return len(set.set)
}

func (set Set) Slice() []string {
	slice := make([]string, set.Len())
	i := 0
	for str := range set.set {
		slice[i] = str
		i++
	}
	return slice
}

func (set Set) String() string {
	strs := []string{}
	for str := range set.set {
		strs = append(strs, fmt.Sprintf("%#v", str))
	}
	return "{" + strings.Join(strs, ", ") + "}"
}

func Equal(s1, s2 Set) bool {
	if s1.Len() != s2.Len() {
		return false
	}
	sl1 := s1.Slice()
	sl2 := s2.Slice()
	sort.Strings(sl1)
	sort.Strings(sl2)
	for i := range sl1 {
		if sl1[i] != sl2[i] {
			return false
		}
	}
	return true
}

func Subset(s1, s2 Set) bool {
	if s1.Len() > s2.Len() {
		return false
	}
	for str := range s1.set {
		if !s2.Has(str) {
			return false
		}
	}
	return true
}

func Disjoint(s1, s2 Set) bool {
	for str := range s1.set {
		if s2.Has(str) {
			return false
		}
	}
	return true
}

func Intersection(s1, s2 Set) Set {
	set := New()
	for str := range s1.set {
		if s2.Has(str) {
			set.set[str] = true
		}
	}
	for str := range s2.set {
		if s1.Has(str) {
			set.set[str] = true
		}
	}
	return set
}

func Union(s1, s2 Set) Set {
	set := New()
	for str := range s1.set {
		set.set[str] = true
	}
	for str := range s2.set {
		set.set[str] = true
	}
	return set
}

func Difference(s1, s2 Set) Set {
	set := New()
	for str := range s1.set {
		if !s2.Has(str) {
			set.set[str] = true
		}
	}
	return set
}

func SymmetricDifference(s1, s2 Set) Set {
	return Union(Difference(s1, s2), Difference(s2, s1))
}
